WalletWasabi/Stores/IndexStore.cs

```diff
@@ -192,204 +120,107 @@ private bool TryProcessFilterNoLock(FilterModel filter, bool enqueue)
+				bool success;
 
-				using (await MatureIndexAsyncLock.LockAsync(CancellationToken.None).ConfigureAwait(false))
-				using (await ImmatureIndexAsyncLock.LockAsync(CancellationToken.None).ConfigureAwait(false))
+				using (await IndexLock.LockAsync(CancellationToken.None).ConfigureAwait(false))
 				{
-					ImmatureIndexFileManager.DeleteMe();
-					MatureIndexFileManager.DeleteMe();
+					success = TryProcessFilterNoLock(filter, enqueue: true);
 				}
 
-				Environment.Exit(2);
+				if (success)
+				{
+					NewFilter?.Invoke(this, filter); // Event always outside the lock.
+				}
 			}
 		}
+	}
 
-		while (ImmatureFilters.Any())
-		{
-			removed.Add(await RemoveLastFilterAsync().ConfigureAwait(false));
-		}
-
-		return removed;
+	public Task<FilterModel?> TryRemoveLastFilterAsync()
+	{
+		return TryRemoveLastFilterIfNewerThanAsync(height: null);
 	}
 
-	/// <summary>
-	/// It'll LogError the exceptions.
-	/// If cancelled, it'll LogTrace the exception.
-	/// </summary>
-	private async Task TryCommitToFileAsync(TimeSpan throttle)
+	private async Task<FilterModel?> TryRemoveLastFilterIfNewerThanAsync(uint? height)
 	{
-		try
+		FilterModel? filter;
+
+		using (await IndexLock.LockAsync(CancellationToken.None).ConfigureAwait(false))
 		{
-			// If throttle is requested, then throttle.
-			if (throttle != TimeSpan.Zero)
+			if (height is null)
 			{
-				// Increment the throttle ID and remember the incremented value.
-				int incremented = Interlocked.Increment(ref _throttleId);
-
-				if (incremented < 21)
+				if (!IndexStorage.TryRemoveLast(out filter))
 				{
-					await Task.Delay(throttle, CancellationToken.None).ConfigureAwait(false);
-				}
-
-				// If the _throttleId is still the incremented value, then I am the latest CommitToFileAsync request.
-				//	In this case I want to make the _throttledId 0 and go ahead and do the writeline.
-				// If the _throttledId is not the incremented value anymore then I am not the latest request here,
-				//	So just return, the latest request will do the file write in its own time.
-				if (Interlocked.CompareExchange(ref _throttleId, 0, incremented) != incremented)
-				{
-					return;
+					throw new InvalidOperationException("No last filter.");
 				}
 			}
 			else
 			{
-				Interlocked.Exchange(ref _throttleId, 0); // So to notify the currently throttled threads that they do not have to run.
+				if (!IndexStorage.TryRemoveLastIfNewerThan(height.Value, out filter))
+				{
+					throw new InvalidOperationException("No last filter.");
+				}
 			}
 
-			using (await IndexLock.LockAsync(CancellationToken.None).ConfigureAwait(false))
-			using (await MatureIndexAsyncLock.LockAsync(CancellationToken.None).ConfigureAwait(false))
-			using (await ImmatureIndexAsyncLock.LockAsync(CancellationToken.None).ConfigureAwait(false))
+			if (SmartHeaderChain.TipHeight != filter.Header.Height)
```

Comment:
Have you considered removing `SmartHeaderChain` and adding that to the DB? It looks to me like `SmartHeaderChain` is trying to solve the fact we cannot get the filters one by one.